pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,234 @@
+// Package scheduler provides a background task scheduler that allows for the registration,
+// execution, and management of periodic tasks. Tasks can be grouped, named, and configured
+// with various options such as custom intervals, log fields, and skip conditions.
+//
+// The scheduler supports dynamic interval updates and can gracefully stop tasks either
+// individually or by group.
+package scheduler
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	cometbft "github.com/cometbft/cometbft/types"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+
+	"github.com/zeta-chain/node/pkg/bg"
+	"github.com/zeta-chain/node/pkg/ticker"
+)
+
+// Scheduler represents background task scheduler.
+type Scheduler struct {
+	definitions map[uuid.UUID]*Definition
+	mu          sync.RWMutex
+	logger      zerolog.Logger
+}
+
+// Task represents scheduler's task
+type Task func(ctx context.Context) error
+
+// Group represents Definition group.
+// Definitions can be grouped for easier management.
+type Group string
+
+// DefaultGroup is the default group for definitions.
+const DefaultGroup = Group("default")
+
+// Definition represents a configuration of a Task
+type Definition struct {
+	// ref to the Scheduler is required
+	scheduler *Scheduler
+
+	// naming stuff
+	id    uuid.UUID
+	group Group
+	name  string
+
+	// arbitrary function that will be invoked by the scheduler
+	task Task
+
+	// represents interval ticker and its options
+	ticker          *ticker.Ticker
+	interval        time.Duration
+	intervalUpdater func() time.Duration
+	skipper         func() bool
+
+	// zeta block ticker (also supports skipper)
+	blockChan       <-chan cometbft.EventDataNewBlock
+	blockChanTicker *blockTicker
+
+	// logging
+	logFields map[string]any
+	logger    zerolog.Logger
+}
+
+// New Scheduler instance.
+func New(logger zerolog.Logger) *Scheduler {
+	return &Scheduler{
+		definitions: make(map[uuid.UUID]*Definition),
+		logger:      logger.With().Str("module", "scheduler").Logger(),
+	}
+}
+
+// Register registers and starts new task in the background
```

Comment:
Can we explain the returned value? Definition is configuration of a Task, to me a configuration is something that would be given as an input